File: backend/app/schemas/architect.py
from typing import Annotated

from pydantic import Field

from app.schemas.base import Schema

ShortText = Annotated[str, Field(min_length=1, max_length=240)]
LongText = Annotated[str, Field(min_length=1, max_length=1200)]


class ArchitectRequest(Schema):
    mission: Annotated[str, Field(min_length=10, max_length=2000)]


class OrganizationDna(Schema):
    collaboration: Annotated[int, Field(ge=0, le=100)]
    creativity: Annotated[int, Field(ge=0, le=100)]
    quality: Annotated[int, Field(ge=0, le=100)]
    security: Annotated[int, Field(ge=0, le=100)]
    speed: Annotated[int, Field(ge=0, le=100)]


class OrganizationRole(Schema):
    name: ShortText
    responsibility: LongText
    worker_count: Annotated[int, Field(ge=1, le=12)]


class OrganizationDepartment(Schema):
    mandate: LongText
    name: ShortText
    roles: Annotated[list[OrganizationRole], Field(min_length=1, max_length=8)]


class OrganizationBlueprint(Schema):
    confidence_score: Annotated[int, Field(ge=0, le=100)]
    departments: Annotated[list[OrganizationDepartment], Field(min_length=1, max_length=8)]
    deliverables: Annotated[list[ShortText], Field(min_length=1, max_length=8)]
    dna: OrganizationDna
    estimated_duration: ShortText
    execution_strategy: LongText
    mission_summary: LongText
    organization_name: ShortText
    organization_type: ShortText
    risks: Annotated[list[ShortText], Field(min_length=1, max_length=8)]
    suggested_culture: ShortText
    worker_capacity: Annotated[int, Field(ge=1, le=64)]

File: frontend/lib/api/architect.ts
export type OrganizationBlueprint = {
  confidence_score: number;
  departments: Array<{
    mandate: string;
    name: string;
    roles: Array<{
      name: string;
      responsibility: string;
      worker_count: number;
    }>;
  }>;
  deliverables: string[];
  dna: {
    collaboration: number;
    creativity: number;
    quality: number;
    security: number;
    speed: number;
  };
  estimated_duration: string;
  execution_strategy: string;
  mission_summary: string;
  organization_name: string;
  organization_type: string;
  risks: string[];
  suggested_culture: string;
  worker_capacity: number;
};

type ApiErrorPayload = { code?: string; message?: string };

export class ArchitectApiError extends Error {
  readonly code: string;
  readonly status: number;

  constructor({
    code,
    message,
    status,
  }: Readonly<{ code: string; message: string; status: number }>) {
    super(message);
    this.name = "ArchitectApiError";
    this.code = code;
    this.status = status;
  }
}

const apiBaseUrl = process.env.NEXT_PUBLIC_API_BASE_URL ?? "http://localhost:8000";

export async function requestOrganizationBlueprint(
  mission: string,
): Promise<OrganizationBlueprint> {
  let response: Response;

  try {
    response = await fetch(`${apiBaseUrl}/api/v1/architect`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ mission }),
    });
  } catch {
    throw new ArchitectApiError({
      code: "organization_architect_unreachable",
      message: "Genesis could not reach the Organization Architect. Please try again.",
      status: 0,
    });
  }

  if (!response.ok) {
    const payload = (await response.json().catch(() => ({}))) as ApiErrorPayload;
    throw new ArchitectApiError({
      code: payload.code ?? "organization_architect_error",
      message: payload.message ?? "Genesis could not create an organization blueprint.",
      status: response.status,
    });
  }

  return (await response.json()) as OrganizationBlueprint;
}
